src/exo/range_analysis.py

```diff
@@ -1,107 +1,289 @@
-            return None
+        if not expr.type.is_indexable():
+            return (None, None)
 
-        if isinstance(e, LoopIR.Read):
-            sym = e.name
-            if sym in self._e_symbols:
+        if isinstance(expr, LoopIR.Read):
+            sym = expr.name
+            if sym in e_symbols:
                 # It is unclear how to do range analysis when a symbol
                 # is read twice within an expression. In most cases,
                 # this won't matter since the expression are normalized
                 # before we try to do range analysis on them
-                return None
-            self._e_symbols.add(sym)
-            return self._env.get(sym)
-        elif isinstance(e, LoopIR.Const):
-            return (e.val, e.val)
-        elif isinstance(e, LoopIR.USub):
-            e_range = self._analyze_range(e.arg)
-            if e_range is None:
-                return None
-            return (-e_range[1], -e_range[0])
-        elif isinstance(e, LoopIR.BinOp):
-            lhs_range = self._analyze_range(e.lhs)
-            rhs_range = self._analyze_range(e.rhs)
-            if lhs_range is None or rhs_range is None:
-                return None
+                return (None, None)
+            e_symbols.add(sym)
+            assert sym in env
+            return env[sym]
+        elif isinstance(expr, LoopIR.Const):
+            return (expr.val, expr.val)
+        elif isinstance(expr, LoopIR.USub):
+            e_range = analyze_range(expr.arg)
+            negate = lambda x: -x if x is not None else x
+            return (negate(e_range[1]), negate(e_range[0]))
+        elif isinstance(expr, LoopIR.BinOp):
+            lhs_range = analyze_range(expr.lhs)
+            rhs_range = analyze_range(expr.rhs)
             merge_binop = {
-                "+": IndexRangeAnalysis.merge_add,
-                "-": IndexRangeAnalysis.merge_sub,
-                "*": IndexRangeAnalysis.merge_mul,
-                "/": IndexRangeAnalysis.merge_div,
-                "%": IndexRangeAnalysis.merge_mod,
+                "+": merge_add,
+                "-": merge_sub,
+                "*": merge_mul,
+                "/": merge_div,
+                "%": merge_mod,
             }
-            return merge_binop[e.op](lhs_range, rhs_range)
+            binop_range = merge_binop[expr.op](lhs_range, rhs_range)
+            return binop_range
         else:
             return None
+
+    return analyze_range(expr)
+
+
+def arg_range_analysis(proc, arg):
+    """
+    Try to find a bounding range on the arguments
+    of a proc.
+
+    Returns: A tuple `T` of length 2 s.t. it represents the range
+    `[T[0], T[1]]` (both inclusive).
+    If `T[0]` or `T[1]` is `None` it represents no knowledge
+    of the value of that side of the range.
+
+    NOTE: This is analysis can be slow. We should
+    consider doing it when constructing a new proc
+    and storing the result within LoopIR.fnarg.
```

Comment:
It might be really important to fix this before merging this PR. Take a look at these numbers:
Before:
Exo tests: 108.20s
BLAS Compilation: 
real    1m58.797s
user    9m6.859s
sys     0m17.264s

After:
Exo tests: 145.80s
BLAS compilation:
real    2m48.349s
user    11m41.682s
sys     0m19.669s